### PR TITLE
include/drivers/flash: document on the unrestricted alignment of reads

### LINKED ISSUE #16439: flash: unify read alignment requirements
### Problem to be covered:
Flash read API implementation have different requirement regards read-out offset and read-out length. These requirements are not precised by any API description.

Most of drivers allow unaligned read,. One driver enforce aligned access in terms on offset and length. One driver requires aligned offset only.

Problem with this diverse behavior cause unwanted complication in upper layer in order to adapt to what any flash driver supports. It cause for instance issue described here: #16438

### Solution:
Sufficient: At last make each driver capable to read unaligned length of data. Better if unaligned offset will be ensured for any driver.

#### Current state of flash drivers `read access requirement`:
(pleas correct here if I missed something)

QMSI has requirement for 4 B alignment of offset and length.
NIOS2 has requirement for 4 B alignment of offset.
NRF, MCUX, Geco, STM, SAM, SPI-NOR, w25qxxdv are capable of unaligned readout

### PR DESCRIPTION
**Pull request by nvlsianpu**
Unaligned read-out capability become fact among all drivers.
Let's cut this in stone as API requirement.

fixes #16439

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

### CODE REVIEW

File: include/drivers/flash.h [hunk 1]
```diff
@@ -84,9 +84,8 @@ __subsystem struct flash_driver_api {
 /**
  *  @brief  Read data from flash
  *
- *  Most of flash drivers support unaligned flash access, but some have
- *  restrictions on the read offset or/and the read size. Please refer to
- *  the driver implementation to get details on the read alignment requirement.
+ *  Each of flash drivers support unaligned flash access, without any
```
Comment (carlescufi) [inline include/drivers/flash.h]: s/Each of/ All

**Review by carlescufi — CHANGES_REQUESTED** [review]

**Review by erwango — APPROVED** [review]

File: include/drivers/flash.h (hunk 1 above)
Comment (nvlsianpu) [reply include/drivers/flash.h]: fixed

**Review by kamlaz — APPROVED** [review]

File: include/drivers/flash.h [hunk 2]
```diff
@@ -84,9 +84,8 @@ __subsystem struct flash_driver_api {
 /**
  *  @brief  Read data from flash
  *
- *  Most of flash drivers support unaligned flash access, but some have
- *  restrictions on the read offset or/and the read size. Please refer to
- *  the driver implementation to get details on the read alignment requirement.
+ *  All flash drivers support unaligned flash access, without any
+ *  restrictions on the read offset and the read size.
```
Comment (anangl) [inline include/drivers/flash.h]: And what with the nrf_qspi_nor driver? It still [requires](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/flash/nrf_qspi_nor.c#L487) the size to be a multiple of 4. Moreover, according to the [nRF QSPI documentation](https://infocenter.nordicsemi.com/topic/ps_nrf52840/qspi.html?cp=4_0_0_5_18_2#register.READ.SRC), the read offset needs to be word-aligned, but the driver does not check it nor seems to be handling it in any other way.

**Review by anangl — CHANGES_REQUESTED** [review]

File: include/drivers/flash.h (hunk 2 above)
Comment (nvlsianpu) [reply include/drivers/flash.h]: ^^ can this driver be aligned to above expectation?

Comment (kamlaz) [reply include/drivers/flash.h]: It has be aligned if this PR will be merged.

Comment (nvlsianpu) [reply include/drivers/flash.h]: @kamlaz can you post such a patch?

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
Blocked by #25292

**Comment by carlescufi** [comment]
API Meeting:

- @nvlsianpu did you check all the current drivers beyond the qspi?
- @nvlsianpu did you consider a compile-time option? I seem to recall you did, but I don't remember what the outcome of that dicussion was
- During the API meeting it was suggested we could default to not supporting unaligned access unless a Kconfig option is enabled, hence the previous question

This will be scheduled for discussion in the next API meeting.

**Comment by pabigot** [comment]
> Blocked by #25292

I don't think that PR blocks this change, but it highlights that QSPI needs work before the new requirement can be confirmed.  That is a minimal change for a very specific function to support using an external flash to store the alternative image for mcuboot (fixing #25289).

Also we should extend the promise ofs to not only alignment but memory region: i.e. that it is permitted to writes of data that is located in flash, rather than RAM, which impacts any implementation (like QSPI) that uses DMA that only works on RAM addresses.

**Comment by nvlsianpu** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/26772 going to add unrestricted alignment of reads for nordic qspi flash driver.

**Review by pabigot — CHANGES_REQUESTED** [review]
I believe the exception for QSPI has been addressed, so this can be updated and merged after other issues addressed.

We should also document the expectations regarding memory region for material written to the device.  In particular that even if a device only supports DMA, and DMA requires the source buffer be in RAM, that it must still be permitted to write material where the source buffer is in flash.  Or, if that is not required, document that on some devices such writes may fail and will return a particular error code.

**Review by anangl — APPROVED** [review]

**Comment by nvlsianpu** [comment]
@pabigot 
> We should also document the expectations regarding memory region for material written to the device. In particular that even if a device only supports DMA, and DMA requires the source buffer be in RAM, that it must still be permitted to write material where the source buffer is in flash. Or, if that is not required, document that on some devices such writes may fail and will return a particular error code.

For being consistent I prefer to force all to support source buffer anywhere.

**Comment by nvlsianpu** [comment]
@pabigot Added note on write source buffer location.

**Comment by carlescufi** [comment]
> @pabigot Added note on write source buffer location.

Thanks @nvlsianpu. @pabigot please take another look.

File: include/drivers/flash.h [hunk 3]
```diff
@@ -121,6 +120,8 @@ static inline int z_impl_flash_read(struct device *dev, off_t offset, void *data
 /**
  *  @brief  Write buffer into flash memory.
  *
+ *  All flash drivers support source buffer located either in RAM or SoC flash.
```
Comment (pabigot) [inline include/drivers/flash.h:123]: For complete clarity this should include the same text on alignment and sizes that was added for read.

**Review by pabigot — CHANGES_REQUESTED** [review]

File: include/drivers/flash.h (hunk 3 above)
Comment (nvlsianpu) [reply include/drivers/flash.h:123]: check one more, propose redaction as I'm not english native

Comment (pabigot) [reply include/drivers/flash.h:123]: How about for read:

> All flash drivers support reads without alignment restrictions on the read offset, the read size, or the destination address.

and for write:

> All flash drivers support a source buffer located either in RAM or SoC flash, without alignment restrictions on the source address, or write size or offset.

?

**Review by pabigot — APPROVED** [review]

**Pull request merged**